Ulduar: Convert Flame Leviathan to BossAI & Spell Lists

insunaa opened the pull request:
## 🍰 Pullrequest
<!-- Describe the Pullrequest. -->
This PR converts Flame Leviathan to BossAI and adds Spell Lists

### Todo / Checklist
<!-- In case some parts are still missing, important notes, breaking changes and other notable items, list them here. -->
- [x] Add SQL
- [x] Fix Turret respawn bug
- [x] Fix Hardmode NPCs not despawning

```sql
DELETE FROM `creature_spell_list` WHERE `Id`=3311300;
INSERT INTO `creature_spell_list` (`Id`, `Position`, `SpellId`, `InitialMin`, `InitialMax`, `RepeatMin`, `RepeatMax`, `Flags`, `TargetId`, `ScriptId`, `Availability`, `Probability`, `Comments`) VALUES
(3311300, 0, 62376, 10000, 10000,  1000,  1000, 0, 1, 0, 100, 0, 'Flame Leviathan - Battering Ram'),
(3311300, 1, 62396, 30000, 30000, 15000, 25000, 0, 0, 0, 100, 0, 'Flame Leviathan - Flame Vents'),
(3311300, 2, 62400,  1000,  1000,  1000,  2000, 0, 0, 0, 100, 0, 'Flame Leviathan - Missile Barrage'),
(3311300, 3, 62374,  1000,  1000, 30000, 30000, 0, 0, 0, 100, 0, 'Flame Leviathan - Pursue'),
(3311300, 4, 62375, 10000, 10000, 10000, 10000, 0, 0, 0, 100, 0, 'Flame Leviathan - Gathering Speed');

DELETE FROM `creature_spell_list_entry` WHERE `Id`=3311300;
INSERT INTO `creature_spell_list_entry` VALUES
(3311300, 'Ulduar - Flame Leviathan', 100, 100);

UPDATE `creature_template` SET `SpellList`=3311300 WHERE `entry`=33113;
UPDATE `creature_template` SET `SpellList`=3311300 WHERE `entry`=34003;

DELETE FROM `creature_spell_list` WHERE (`Id` = 3314200) AND (`Position` = 0);
INSERT INTO `creature_spell_list` (`Id`, `Position`, `SpellId`, `InitialMin`, `InitialMax`, `RepeatMin`, `RepeatMax`, `Flags`, `TargetId`, `ScriptId`, `Availability`, `Probability`, `Comments`) VALUES
(3314200, 0, 62402, 2000, 2000, 2000, 4000, 2, 100, 0, 100, 0, 'Leviathan Defense Turret - Searing Flame');

DELETE FROM `creature_spell_list_entry` WHERE `Id`=3314200;
INSERT INTO `creature_spell_list_entry` VALUES
(3314200, 'Ulduar - Leviathan Defense Turret', 100, 100);

UPDATE `creature_template` SET `SpellList`=3314200 WHERE `entry`=33142;

DELETE FROM `creature_spell_list` WHERE (`Id` = 3306200);
INSERT INTO `creature_spell_list` (`Id`, `Position`, `SpellId`, `InitialMin`, `InitialMax`, `RepeatMin`, `RepeatMax`, `Flags`, `TargetId`, `ScriptId`, `Availability`, `Probability`, `Comments`) VALUES
(3306200, 0, 62974, 0, 0, 0,     0,     0, 0, 0, 100, 0, 'Salvaged Chopper - Sonic Horn'),
(3306200, 1, 62286, 0, 0, 15000, 15000, 0, 0, 0, 100, 0, 'Salvaged Chopper - Tar'),
(3306200, 2, 62299, 0, 0, 0,     0,     0, 0, 0, 100, 0, 'Salvaged Chopper - Speed Boost'),
(3306200, 3, 67372, 0, 0, 0,     0,     0, 0, 0, 100, 0, 'Salvaged Chopper - Grab Pyrite'),
(3306200, 5, 64660, 0, 0, 60000, 60000, 0, 0, 0, 100, 0, 'Salvaged Chopper - First Aid Kit');

UPDATE `creature_template` SET `SpellList`=3306200 WHERE `Entry`=33062;

UPDATE `creature_template` SET `ScriptName`='npc_salvaged_chopper' WHERE `Entry`=33062;
UPDATE `creature_template` SET `ScriptName`='npc_pyrite_safety_container' WHERE `Entry`=33218;
UPDATE `creature_template` SET `ScriptName`='npc_liquid_pyrite' WHERE `Entry`=33189;

UPDATE `vehicle_accessory` SET `seat`=0 WHERE `vehicle_entry` IN (33364, 33369, 33108, 33366);

DELETE FROM `spell_template` WHERE `Id`=62474;
INSERT INTO `spell_template` (Id, Attributes, AttributesEx, AttributesEx3, AttributesEx4, AttributesEx6, CastingTimeIndex, MaxLevel, BaseLevel, SpellLevel, RangeIndex, Effect1, EffectDieSides1, EffectBasePoints1, EffectImplicitTargetA1, EffectMultipleValue1, EffectMiscValue1, SpellIconID, SpellName, SpellFamilyFlags) VALUES
(62474, 553648256, 1024, 262144, 128, 16777216, 1, 83, 80, 80, 12, 62, 1, 4, 97, 1, 3, 3779, '(Serverside/Non-DB2) Hurl Boulder (Power Burn)', 16);
DELETE FROM `spell_template` WHERE `Id`=64998;
INSERT INTO `spell_template` (Id, Attributes, AttributesEx, AttributesEx3, AttributesEx4, AttributesEx6, CastingTimeIndex, MaxLevel, BaseLevel, SpellLevel, RangeIndex, Effect1, Effect2, EffectImplicitTargetA1, EffectImplicitTargetA2, SpellName, SpellFamilyFlags, IsServerSide) VALUES
(64998, 553648256, 1024, 262144, 128, 16777216, 1, 83, 80, 80, 1, 3, 3, 94, 1, '(Serverside/Non-DB2) Say you''re ready to fly!', 16, 1);
DELETE FROM `spell_template` WHERE `Id`=62543;
INSERT INTO `spell_template` (Id, Attributes, CastingTimeIndex, DurationIndex, RangeIndex, Effect1, EffectDieSides1, EffectImplicitTargetA1, EffectRadiusIndex1, EffectMiscValue1, EffectMiscValueB1, SpellVisual, SpellIconID, SpellName, SchoolMask, IsServerSide) VALUES
(62543, 384, 1, 9, 1, 28, 1, 22, 36, 33189, 64, 215, 2651, '(Serverside/Non-DB2) Spawn Pyrite', 1, 1);

UPDATE `creature_template` SET `UnitFlags`=`UnitFlags`|33554432 WHERE `entry`=33142;
UPDATE `creature_template` SET `StaticFlags1`=`StaticFlags1`|512 WHERE `entry`=33142;

UPDATE `creature_template` SET `DisplayIdProbability1`=100,`DisplayIdProbability2`=0,`DisplayIdProbability3`=0,`DisplayIdProbability4`=0 WHERE `entry` IN (33189, 33114, 33167);
UPDATE `creature_template` SET `DisplayIdProbability1`=0,`DisplayIdProbability2`=100,`DisplayIdProbability3`=0,`DisplayIdProbability4`=0 WHERE `entry` IN (33264);

UPDATE `creature_ai_scripts` SET `action1_type`=11,`action1_param1`=62987,`action1_param3`=2 WHERE `id`=3321401;

UPDATE `creature_template` SET `InhabitType`=1 WHERE entry=33218;

UPDATE `broadcast_text` SET `ChatTypeID`=3 WHERE `Id`=33746;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15506 WHERE `Id`=33487;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15521 WHERE `Id`=33507;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15520 WHERE `Id`=33506;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15507 WHERE `Id`=33488;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15508 WHERE `Id`=33489;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15509 WHERE `Id`=33490;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15516 WHERE `Id`=33501;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15517 WHERE `Id`=33503;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15518 WHERE `Id`=33504;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15519 WHERE `Id`=33505;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15510 WHERE `Id`=33491;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15512 WHERE `Id`=33493;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15513 WHERE `Id`=33495;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15515 WHERE `Id`=33499;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15514 WHERE `Id`=33497;
UPDATE `broadcast_text` SET `ChatTypeID`=1,`SoundEntriesID1`=15511 WHERE `Id`=33492;
UPDATE `broadcast_text` SET `ChatTypeID`=3 WHERE `Id`=33502;
UPDATE `broadcast_text` SET `ChatTypeID`=3 WHERE `Id`=33494;
UPDATE `broadcast_text` SET `ChatTypeID`=3 WHERE `Id`=33498;
UPDATE `broadcast_text` SET `ChatTypeID`=3 WHERE `Id`=33496;
UPDATE `broadcast_text` SET `ChatTypeID`=3 WHERE `Id`=33500;

DELETE FROM `string_id` WHERE `Id` IN (6030001, 6030002);
INSERT INTO `string_id` (`Id`, `Name`) VALUES
(6030001,'ULDUAR_PLAYER_VEHICLES'),
(6030002,'ULDUAR_FLAME_LEVIATHAN_PARTY');

UPDATE `creature_template` SET `StringId1`=6030001 WHERE `entry` IN (33060, 33062, 33067, 33109, 33167);
UPDATE `creature_template` SET `StringId1`=6030002 WHERE `entry` IN (33113, 33114, 33139, 33142, 33143, 34286);

--UPDATE `creature_template` SET `VisibilityDistanceType`=4 WHERE `entry`=33113;

UPDATE `creature_template` SET `VehicleTemplateId`=340 WHERE `entry`=34003;

DELETE FROM `vehicle_accessory` WHERE `vehicle_entry`=33113 AND `seat` IN (0, 1);
INSERT INTO `vehicle_accessory` VALUES
(33113,0,33114,'Flame Leviathan(25)'),
(33113,1,33114,'Flame Leviathan(25)');

DELETE FROM `spell_script_target` WHERE `entry` IN (62323, 62399);
INSERT INTO `spell_script_target` VALUES (62323, 1, 33114, 6);
INSERT INTO `spell_script_target` VALUES (62399, 1, 33113, 0);
INSERT INTO `spell_script_target` VALUES (62399, 1, 33139, 6);

UPDATE `spell_template` SET `ChannelInterruptFlags`=0 WHERE Id=62399;
UPDATE `creature_template` SET `RegenerateStats`=12 WHERE entry=33142;
UPDATE `creature_template` SET `ScriptName`='npc_mimiron_inferno' WHERE `entry`=33369;
UPDATE `creature_template` SET `ScriptName`='npc_hodir_fury' WHERE `entry`=33212;
UPDATE `creature_template` SET `ScriptName`='npc_hodir_fury_reticle' WHERE `entry`=33108;
UPDATE `creature_template` SET `ScriptName`='npc_freya_ward' WHERE `entry`=33366;
UPDATE `creature_template` SET `ScriptName`='npc_salvaged_demolisher' WHERE `entry`=33109;
UPDATE `creature_template` SET `ScriptName`='npc_leviathan_defense_turret',`UnitFlags`=`UnitFlags`-33554432 WHERE `entry`=33142;

INSERT INTO `creature_ai_scripts` (`id`, `creature_id`, `event_type`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `action1_type`, `comment`) VALUES
(3313903, 33139, 30, 100, 6, 5, 33113, 38, 'Flame Leviathan Turret - Set Turrent in Combat with Zone on Receive AI_EVENT_CUSTOM_EVENTAI_A');

-- for testing purposes only:

DELETE FROM `spell_scripts` WHERE `Id` IN (62482, 62374, 62297, 62323, 62324, 62907, 62910, 62912, 62336, 64414, 62399, 62475, 60603, 63575, 63605, 63618, 61242, 64979, 64998, 65045, 60075, 60076, 60077);
INSERT INTO `spell_scripts` VALUES
(62482,'spell_grab_crate_leviathan'),
(62374,'spell_pursue_leviathan'),
(62297,'spell_hodirs_fury_leviathan'),
(62323,'spell_hookshot'),
(62324,'spell_throw_passenger'),
(62907,'spell_freyas_ward_leviathan'),
(62910,'spell_mimirons_inferno_leviathan'),
(62912,'spell_thorims_hammer_leviathan'),
(62336,'spell_hookshot_aura'),
(64414,'spell_load_into_catapult_leviathan'),
(62399,'spell_overload_circuit'),
(62475,'spell_systems_shutdown'),
(60603,'spell_eject_passenger_1'),
(63575,'spell_smoke_trail_leviathan'),
(63605,'spell_rope_beam'),
(63618,'spell_overload_leviathan'),
(61242,'spell_parachute_leviathan'),
(64979,'spell_anti_air_rocket'),
(62363,'spell_anti_air_rocket'),
(62634,'spell_demolisher_mortar'),
(62635,'spell_demolisher_mortar'),
(64998,'spell_ready_to_fly'),
(65044,'spell_flames_leviathan'),
(65045,'spell_flames_leviathan'),
(60075,'spell_flame_leviathan_buff'),
(60076,'spell_flame_leviathan_buff'),
(60077,'spell_flame_leviathan_buff');
```

Random notes:
```
    /*
Areatrigger left Repair: 5369
Areatrigger right repair: 5423
    */

```

insunaa committed:
Remove before Merge

Add Meson Build System to CMaNGOS WotLK

Meson: Add SQLite

Meson: Update wrap dependencies

Meson: Fix some Windows issues

Meson: Remove Experiment

Meson: Add license, minimum version, fix PCH

Add changes to Log.h

Meson: Remove unused files

Meson: Ignore Modules

Meson: Add Botanica changes

Meson: Set new minimum OpenSSL version

Meson: Update wrap dependencies

Meson: Remove Maker from list

Meson: Add CodeBench

Meson: Add BGQ and Mana Tombs

Meson: Add apple silicon linker flags

Meson: fix missing variable declaration

Meson: Add ability to build new playerbots

Fix CodeBench Non-PCH compile

Add Argent Tournament scripts

Change to C++20 Standard

Add SpellStacking to build system

insunaa committed:
BossAI: Add Queued Casts

BossAI: Add timed respawn on evade

BossAI: Add despawning of subordinates on Evade

BossAI: Add Resettable Values

BossAI: Minor execution order fix

BossAI: Remove resettable values

BossAI: Remove Subordinate-Despawn

BossAI: Change documentation for QueuedCasts

insunaa committed:
Ulduar: Convert Flame Leviathan to BossAI + Spell Lists

Ulduar: Minor Leviathan improvements.
Fix hardmodes not despawning

Ulduar: Get Flame Leviathan to a working state

insunaa committed:
Vehicles/Transport: Fix boarding with existing driver

Vehicles: Also count passengers inside of passengers

Vehicles: More Vehicle fixes

insunaa committed:
Ulduar: Leviathan Fixes

Ulduar: Add Smoke Trail spellscript

Ulduar: Ensure SpellClick is active during Leviathan encounter

Ulduar: Add Spellscript for Parachute

Ulduar: Remove incorrect line, add comment

Ulduar: Unmagic some magic numbers

Ulduar: Fix targeting reticle npc issues

Ulduar: Minor Leviathan fixes

Ulduar: Fix demolisher energy-sync & Targetting issues

Ulduar: Flip seat Power sync

Ulduar: Minor Leviathan fixes

Fix Hodir's Fury Reticle

Ulduar: Add Leashing to Flame Leviathan

Ulduar: fix spellscript for eject passenger 1

Ulduar: Partially fix initial Leviathan Gossip

Ulduar: Fix friendly fire by vehicles

Ulduar: Open Leviathan door on Fail

Ulduar: Make Thorim's Hammer repeatable

Ulduar: Fix Freya-Tower

Ulduar: Fix Hodir's Fury

Ulduar: Fix Thorim's Hammer

Ulduar: Rework Demolisher Energy

Ulduar: Make Bosses immune to Knockback

Ulduar: Implement Ready to Fly spell and fix seat ejections

Ulduar: Minor semantic fix

insunaa commented:
OK I consider this PR as functionally complete.
The following things related to the encounter either don't work or don't work correctly:
- The pyrite helicopter things don't have waypoints
- The shot down pyrite things have the display ID of boxes instead of the blue pyrite containers they should be
- The vehicles aside from the Demolisher are untested
- The demolisher doesn't correctly share its Pyrite/Power meter with the passenger/mechanic seat
- The spell projectiles for the hardmode spells aren't really projectiles, or at least aren't visible
- The defense turrets are not uninteractible for players who are not on top of the leviathan (no clear evidence found for either way, if this mechanic exists it's a deeper core issue)
- The summoned creatures from Freya's hardmode despawn when their spell timer is up, when they should instead stay and attack players.

I don't have the patience to fix these issues right now and I'd appreciate it a lot if someone else would consider helping me :+1:

insunaa committed:
Allow triggering spell to serve as casterauraspell

insunaa committed:
Ulduar: Fully working pyrite refuel

Ulduar: Final(ish) Pyrite Refueling fix

Ulduar: Fix Friendly Fire for AA Rocket

Ulduar: Fix Mechanolifts

Ulduar: Implement Serverside Pyrite Spell

Ulduar: neutral targetting for pyrite spells

Leviathan: Cleanup

insunaa committed:
Ulduar: Improve vehicle spellclick behaviour

insunaa committed:
WotLK: add Eject Passenger (Chopper) SpellScript

insunaa committed:
Ulduar: Spawn dead-by-default vehicles as dead

insunaa committed:
Vehicles: Eject Passengers across subvehicles

insunaa committed:
Ulduar: Fix BCTs for Leviathan

Ulduar: Kill Leviathan Adds on death

Ulduar: Make reticles float for spell visibility

Ulduar: Increase Bubble visbility

Ulduar: Fix Freya's Ward spawn (Levi)

Ulduar: Improve Hookshot Aura and Shutdown BCT

Leviathan: Update SD2 comment

Ulduar: Improve Salvaged Chopper

Ulduar: Cleanup in Leviathan code

Ulduar: Simplify Chopper code slightly

Ulduar: Temporarily disable trigger spell

Ulduar: Use new vehicle hooks

AnonXS commented on src/game/Entities/TemporarySpawn.cpp:
```diff
@@ -174,7 +174,7 @@ void TemporarySpawn::Update(const uint32 diff)
             if (!GetCharmerGuid())
             {
-                if (IsExpired())
+                if (IsExpired() && !IsInCombat())
```
Can confirm this behavior with ai findings related to npc Dread Tactician 16959 using s.33679 (15000 duration), he can have multiple adds up during combat which will be despawned on combat end/reached home.

![grafik](https://user-images.githubusercontent.com/19734826/219691381-079df1a3-5cdf-4bf2-a24f-118c364a2af8.png)

insunaa committed:
Respect Knockback Immunity

insunaa committed:
Spell: Handle Knockback immunity in boss script

insunaa committed:
Vehicles: Ensure vehicle attachments stay attached

insunaa committed:
Ulduar: Use BossAI door controls for Leviathan

insunaa committed:
Spells: Allow secondary target checking for area auras

insunaa committed:
Spells: Allow secondary target checking for area auras

insunaa committed:
Fix Mimiron's Fury

insunaa committed:
Ulduar: Fix AreaAuraCheck for Leviathan

insunaa committed:
Vehicles: Allow switching between vehicles from within vehicles

insunaa committed:
Spells: Add Spellscripts to spell.sql

insunaa committed:
BossAI: Minor Leviathan touchups

BossAI: prepare for rebase

insunaa committed:
Leviathan: Use string_ids

insunaa committed:
BossAI: Only target Flame Leviathan and Friends for hardmode spell

BossAI: cleanup

insunaa committed:
Ulduar: Fix OnAreaAuraCheckTarget invocation

Ulduar: Leviathan optimizations

Ulduar: Update Spellscript data

insunaa committed:
Vehicles: Fix various issues
- Fix passenger sometimes visually not mounting
- Fix wrong seat coordinate if the player is the vehicle
- Deduplicate some code

Vehicles: Fix switching seats

insunaa committed:
Leviathan: Remove debug-print include

Leviathan: Fix door

insunaa committed:
Vehicles: Don't respawn entities in non-existent seats

Vehicles: Remove check for switching vehicles directly

Vehicle: Fix memory leak

insunaa commented on src/game/Entities/Vehicle.cpp:
```diff
@@ -788,11 +796,29 @@ uint8 VehicleInfo::GetTakenSeatsMask() const
     uint8 takenSeatsMask = 0;
 
     for (const auto& m_passenger : m_passengers)
+    {
+        if (m_passenger.first->IsUnit())
+            if (static_cast<Unit*>(m_passenger.first)->IsVehicle() && static_cast<Unit*>(m_passenger.first)->GetVehicleInfo()->GetEmptySeats() > 0)
+                continue;
         takenSeatsMask |= 1 << m_passenger.second->GetTransportSeat();
+    }
```
I don't know what I was thinking here, need to fix this

insunaa commented on src/game/Entities/VehicleHandler.cpp:
```diff
@@ -42,12 +42,20 @@ void WorldSession::HandleDismissControlledVehicle(WorldPacket& recvPacket)
 
     Unit* vehicle = (Unit*)transportInfo->GetTransport();
 
+    if (!vehicle)
+        return;
     // Something went wrong
     if (vehicleGuid != vehicle->GetObjectGuid())
         return;
+    if (vehicle->IsBoarded())
+    {
+        vehicle = static_cast<Unit*>(vehicle->GetTransportInfo()->GetTransport());
+    }
 
     // Remove Vehicle Control Aura
     vehicle->RemoveSpellsCausingAura(SPELL_AURA_CONTROL_VEHICLE, _player->GetObjectGuid());
+
+    vehicle->GetVehicleInfo()->UnBoard(_player, false);
```
not sure what I was thinking here either. unboarding from the parent vehicle for some reason?

insunaa replied:
unboarding works recursively for subvehicles, so it's probably fine and possibly cleans up misapplied boardings in parent vehicles

insunaa commented on src/game/Maps/TransportSystem.cpp:
```diff
@@ -154,21 +155,43 @@ bool TransportBase::HasOnBoard(WorldObject const* passenger) const
 
+    for (const auto& m_passenger : m_passengers)
+    {
+        if (transportInfo->GetTransportSeat() == m_passenger.second->GetTransportSeat())
+            if (m_passenger.first->IsUnit())
+            {
+                if (static_cast<const Unit*>(m_passenger.first)->IsVehicle())
+                {
+                    static_cast<const Unit*>(m_passenger.first)->GetVehicleInfo()->Board(static_cast<Unit*>(passenger), 0);
+                    return false;
+                }
+            }
```
style issue

insunaa commented on src/game/Spells/Spell.cpp:
```diff
@@ -7009,7 +7011,7 @@ SpellCastResult Spell::CheckCast(bool strict)
                         if (!boardedOnEachOther)
                             return SPELL_FAILED_NOT_ON_TRANSPORT;
-                    }
+                    }*/
```
needs some code analysis. the code does the exact opposite of what it is supposed to do, so I commented it out. but I haven't investigated the mechanism for why it's wrong, yet.

insunaa commented on src/game/Spells/Spell.cpp:
```diff
@@ -4088,7 +4088,8 @@ void Spell::update(uint32 difftime)
     // check if the player or unit caster has moved before the spell finished (exclude casting on vehicles)
     if ((m_trueCaster->IsUnit() && m_timer != 0) &&
             (m_castPositionX != m_trueCaster->GetPositionX() || m_castPositionY != m_trueCaster->GetPositionY() || m_castPositionZ != m_trueCaster->GetPositionZ()) &&
-            (m_spellInfo->Effect[EFFECT_INDEX_0] != SPELL_EFFECT_STUCK || !m_trueCaster->m_movementInfo.HasMovementFlag(MOVEFLAG_FALLINGFAR)))
+            (m_spellInfo->Effect[EFFECT_INDEX_0] != SPELL_EFFECT_STUCK || !m_trueCaster->m_movementInfo.HasMovementFlag(MOVEFLAG_FALLINGFAR)) &&
+            (!m_trueCaster->m_movementInfo.HasMovementFlag(MOVEFLAG_ONTRANSPORT)))
```
should probably be re-ordered for efficiency falling < transport < stuck

insunaa commented on src/game/Entities/Unit.cpp:
```diff
@@ -9181,6 +9181,12 @@ bool Unit::IsVisibleForOrDetect(Unit const* u, WorldObject const* viewPoint, boo
+            if (Player* player = const_cast<Player*>(dynamic_cast<const Player*>(u)))
+                if (!player->HasAtClient(vehicle))
+                    return false;
+
```
I'm sure I had a more efficient version of this somewhere

insunaa commented on src/game/Spells/Spell.cpp:
```diff
@@ -5610,7 +5610,8 @@ SpellCastResult Spell::CheckCast(bool strict)
         // Caster aura req check if need
         if (m_spellInfo->casterAuraSpell && !m_caster->HasAura(m_spellInfo->casterAuraSpell))
-            return SPELL_FAILED_CASTER_AURASTATE;
+            if (!m_triggeredBySpellInfo || m_triggeredBySpellInfo->Id != m_spellInfo->casterAuraSpell)
+                return SPELL_FAILED_CASTER_AURASTATE;
```
important, but I don't remember what for

insunaa committed:
Ulduar: Remove unneeded Leviathan changes

insunaa committed:
Vehicle: Style Changes

Vehicle: Return unit to passive instead of aggressive after unboarding

insunaa commented:
For easier review I created a new PR on my fork which has the Meson commits stripped and the Leviathan commits squashed: https://github.com/insunaa/mangos-wotlk/pull/1/files

Please someone take the 30 minutes it takes to review. Thank you!

insunaa committed:
Leviathan: Minor style changes

Leviathan: Fix friendly fire and mulitboarding

al3xc1985 commented:
Any news on this one @insunaa, I mean since my last test. Cause this one would fix a lot of problems in game when merged

insunaa commented:
@al3xc1985 I'm catching up on work due to illness right now, so little time for this, hopefully soon.
